src/plugins/visualizations/public/embeddable/to_ast.ts

```diff
@@ -19,31 +18,35 @@ import { VisToExpressionAst } from '../types';
  *
  * @internal
  */
-export const toExpressionAst: VisToExpressionAst = async (vis, params) => {
-  const { savedSearchId, searchSource } = vis.data;
-  const query = searchSource?.getField('query');
-  let filters = searchSource?.getField('filter');
-  if (typeof filters === 'function') {
-    filters = filters();
+export const toExpressionAst: VisToExpressionAst = async (
+  vis,
+  params
+): Promise<ExpressionAstExpression> => {
+  if (!vis.type.toExpressionAst) {
+    throw new Error('Visualization type definition should have toExpressionAst function defined');
   }
 
-  const kibana = buildExpressionFunction<ExpressionFunctionKibana>('kibana', {});
-  const kibanaContext = buildExpressionFunction<ExpressionFunctionKibanaContext>('kibana_context', {
-    q: query && queryToAst(query),
-    filters: filters && filtersToAst(filters),
-    savedSearchId,
-  });
-
-  const ast = buildExpression([kibana, kibanaContext]);
-  const expression = ast.toAst();
+  const searchSource = vis.data.searchSource?.createCopy();
 
-  if (!vis.type.toExpressionAst) {
-    throw new Error('Visualization type definition should have toExpressionAst function defined');
+  if (vis.data.aggs) {
+    vis.data.aggs.hierarchical = vis.isHierarchical();
+    searchSource?.setField('aggs', vis.data.aggs);
   }
 
   const visExpressionAst = await vis.type.toExpressionAst(vis, params);
-  // expand the expression chain with a particular visualization expression chain, if it exists
-  expression.chain.push(...visExpressionAst.chain);
+  const searchSourceExpressionAst = searchSource?.toExpressionAst();
+
+  const expression = {
+    ...visExpressionAst,
+    chain: [
+      buildExpressionFunction<ExpressionFunctionKibana>('kibana', {}).toAst(),
+      /**
+       * @workaround Non-aggregating visualizations only accept the `kibana_context` on input.
+       */
+      ...(searchSourceExpressionAst?.chain.filter(({ function: name }) => name !== 'esdsl') ?? []),
```

Comment:
This seems pretty hacky to me and it also means no visualization will be able to use esdsl. This is done for TSVB and Timelion, right?

Instead of building the esdsl expression, then throwing it away what about a flag on the visualization definition whether it needs the search source ast prepended or whether it will take care of that internally and handling the special case within the TSVB/Timelion `toExpressionAst` function: 

https://github.com/elastic/kibana/blob/3730dd0779ed8efd74aee88f57422781ec1ac122/src/plugins/vis_types/timeseries/public/to_ast.ts#L15

Reply:
@flash1293 That was a great idea. I've updated the PR to add this extra parameter. Could you please take another look?

cc @ppisljar